### PR TITLE
upd/`neo-go` to `v0.102.0`

### PR DESCRIPTION
**Pull request by carpawell**

### CODE REVIEW

File: pkg/morph/deploy/util.go [hunk 1]
```diff
@@ -41,7 +41,9 @@ func isErrNotEnoughGAS(err error) bool {
 }
 
 func isErrInvalidTransaction(err error) bool {
-	return errors.Is(err, neorpc.ErrValidationFailed)
+	// neo-go changed -504 code to be a network fee error,
+	// but it was a verification error before v0.102.0
+	return errors.Is(err, neorpc.ErrInsufficientNetworkFee)
```
Comment (roman-khimov) [inline pkg/morph/deploy/util.go]: That's a straight `isErrNotEnoughGAS`, while `isErrInvalidTransaction` is any of -5xx (you can get a lot of different codes now).

File: pkg/morph/client/notifications.go [hunk 2]
```diff
@@ -107,3 +164,167 @@ func (c *Client) UnsubscribeAll() error {
...
+	curBlockChan  chan *block.Block
+	curNotaryChan chan *result.NotaryRequestEvent
+
+	// cached subscription information
+	subscribedEvents       map[util.Uint160]struct{}
+	subscribedNotaryEvents map[util.Uint160]struct{}
+	subscribedToNewBlocks  bool
+}
+
+func (c *Client) routeNotifications() {
```
Comment (roman-khimov) [inline pkg/morph/client/notifications.go:199]: Why do we still reroute?

File: pkg/morph/client/constructor.go [hunk 3]
```diff
@@ -145,8 +145,10 @@ func New(key *keys.PrivateKey, opts ...Option) (*Client, error) {
 }
 
 func (c *Client) newCli(endpoint string) (*rpcclient.WSClient, *actor.Actor, error) {
-	cli, err := rpcclient.NewWS(c.cfg.ctx, endpoint, rpcclient.Options{
-		DialTimeout: c.cfg.dialTimeout,
+	cli, err := rpcclient.NewWS(c.cfg.ctx, endpoint, rpcclient.WSOptions{
+		Options: rpcclient.Options{
+			DialTimeout: c.cfg.dialTimeout,
```
Comment (carpawell) [inline pkg/morph/client/constructor.go:178]: now sure what do we do with `CloseNotificationChannelIfFull`

Comment (roman-khimov) [reply pkg/morph/client/constructor.go:178]: It's a deadlock safeguard basically, likely should be set. @AnnaShaleva?

Comment (carpawell) [reply pkg/morph/client/constructor.go:187]: > It's a deadlock safeguard basically

yes, but do we need it? it leads to unexpected connection switches, not sure if it is ok for the current logic. but if we want if, i will recheck it

File: pkg/morph/deploy/util.go (hunk 1 above)
Comment (carpawell) [reply pkg/morph/deploy/util.go]: yes, usually i split "big" libraries updates into two (or more) commits: one for the `go.mod` file and the other for the features adoptions. the problem was that `ErrValidationFailed` just does not exist anymore so it could not be complied. that is why i put it here (another -504 code, even if it is named weird for that part of the code). it is dropped right in the next commit

File: pkg/morph/client/constructor.go (hunk 3 above)
Comment (roman-khimov) [reply pkg/morph/client/constructor.go:187]: > it leads to unexpected connection switches...

...or a deadlock. But you can choose which one you like more!

File: pkg/morph/deploy/util.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/morph/deploy/util.go]: Yeah, I've seen that as well.

File: pkg/morph/client/constructor.go (hunk 3 above)
Comment (carpawell) [reply pkg/morph/client/constructor.go:187]: well, not sure what is worse fot debugginh: unexpected connection switch or unexpected deadlock so added a commit that turns that new option on

File: pkg/morph/client/constructor.go [hunk 4]
```diff
@@ -71,10 +71,10 @@ func defaultConfig() *cfg {
 	}
 }
 
-// ErrStaleNode is returned from [New] when minimal required height
+// ErrStaleNodes is returned from [New] when minimal required height
 // requirement specified in [WithMinRequiredBlockHeight] is not
 // satisfied by the given nodes.
-var ErrStaleNode = errors.New("RPC node is not yet up to date")
+var ErrStaleNodes = errors.New("RPC nodes are not yet up to date")
```
Comment (roman-khimov) [inline pkg/morph/client/constructor.go:77]: Why this change, really?

Comment (carpawell) [reply pkg/morph/client/constructor.go:77]: cause now it is retuned only if _every_ node is _not_ up to date. calling `SwitchRPC` outside of the morph pkg is strange to me

**Review by roman-khimov — APPROVED** [review]

File: pkg/innerring/internal/blockchain/blockchain.go [hunk 5]
```diff
@@ -352,12 +352,12 @@ func New(cfg Config) (res *Blockchain, err error) {
 	cfgBaseProto.MaxTraceableBlocks = cfg.TraceableChainLength
 	cfgBaseProto.Hardforks = cfg.HardForks
 	if cfg.ValidatorsHistory != nil {
-		cfgBaseProto.ValidatorsHistory = make(map[uint32]int, len(cfg.ValidatorsHistory))
+		cfgBaseProto.ValidatorsHistory = make(map[uint32]uint32, len(cfg.ValidatorsHistory))
```
Comment (cthulhu-rider) [inline pkg/innerring/internal/blockchain/blockchain.go:355]: no need to copy elements anymore

**Review by cthulhu-rider — APPROVED** [review]
